### PR TITLE
CodeEditorExample

### PR DESCRIPTION
**Pull request by dvd101x**
Hi,

This is an example of a code editor in the browser according to #2719.

I would like to know if the functionality is as expected.

Some features that could be included are:
- save states in local storage
- make more sessions (let's say 10 editors to choose from)
- include comments as markdown in the results
- format the results with latex instead of text
- do the calculations on a web worker

The code could be improved, but first I would like to know if the result is according to your vision and what do you think about the possible features.

### CODE REVIEW

**Comment by dvd101x** [comment]
Thanks 😄,

It makes sense,

I did include some features that could be removed to simplify a bit:

1. It currently separates the input into paragraphs(blocks), thus creating these chunks of results per paragraph.
2. Has a delay between input and results (so that it doesn't evaluate every keystroke)
3. It splits the input in lines that can be parsed together to match the output. It could be reduced significantly by just evaluating the whole input and show the whole set of results. (Also if a single expressions throws an error, the whole input would throw an error)
4. Has some code that gets the examples from certain functions, it could be reduced by including a text at the beginning

What do you think, shall some of those features be removed to be bare-bone?

**Comment by josdejong** [comment]
I think it would indeed simplify the logic related to separate blocks a bit. It would be nice to keep evaluating lines one by one because else a single line having an error would indeed make the whole output an error.

The delay between input and output makes a lot of sense, I think it is important to show that in the example.

I like the autocompletion, it shows a very useful and tight integration between the code editor and mathjs.

I think it would be easier to grasp what is going on when providing a static string with some initial calculations (maybe take the ones from https://mathjs.org/), instead of smartly loading them via the help examples of a list of functions.

**Comment by dvd101x** [comment]
Thanks for your inputs!

The use of decoratros is something that I would be interested in implementing to get rid of the right panel.

Sadly I don't have much experience with **bundlers**, thus I was sticking to CodeMirror5 which doesn't have that option.

If you could point me to a way to make a simple CM6 editor the right way within this project I'm interested in migrating the CM5 logic and including the results in decorators which would be really cool.

**Comment by josdejong** [comment]
I'll give it a try and do some fiddling the end of next week, that would be fun. I have some recent experience with CodeMirror6 in a different project, so I'm not a total stranger.

**Comment by josdejong** [comment]
I made a little start, but didn't get very far today: https://github.com/josdejong/mathjs-codemirror

**Comment by dvd101x** [comment]
Hi, this is very cool, I was able to work a little on this to include syntax highlighting with a legacy mode. Made a PR

**Comment by josdejong** [comment]
The experiment in https://github.com/josdejong/mathjs-codemirror is quite functional now. Showing results inline adds quite some complexity, so for this example it may be best to take your approach of showing the results in a panel on the right side instead, so we keep the example as simple as possible. So then we just copy the setup of the editor to this example and not the logic for inline results.

Also, the language definition should still be converted to CodeMirror 6 if I'm not mistaken, but I'm not sure what is involved in that and how much work it will be.

**Comment by dvd101x** [comment]
Yes, the experiment is fantastic, but maybe not so easy to implement at first.

Regarding the language definition with a non legacy mode I think it's a nice to have, there are many languages that have not made the transition to CodeMirror 6 including Matlab/Octave. A reference for complexity could be Julia that has this [community project](https://github.com/JuliaPluto/lang-julia) but I don't think it should stop this example.

**Comment by dvd101x** [comment]
Hi Jos, thanks to https://github.com/josdejong/mathjs-codemirror I think I can translate this simpler example to run with vite now, should that logic be migrated to use webpack to work in this repository?

**Comment by josdejong** [comment]
I think the example should be standalone, and the mathjs bundler should not also bundle examples. So I think you can create a mini project with vite and a package.json in the examples. Does that make sense?

**Comment by dvd101x** [comment]
ok, it makes sense, I can include the mini project in examples with vite, but I will definitely need help on how to call the build step outside vite and the folder structure for it to work in the website.

I will work on it and report back.

**Comment by josdejong** [comment]
> I will definitely need help on how to call the build step outside vite and the folder structure for it to work in the website.

Hm, that is a good point. I haven't thought that through. I think we can create a script in `package.json` which goes to the example, runs `npm install` and then `npm run build`. Maybe easiest if I do that myself in the `gh-pages` branch since I can deploy and test that.

**Comment by dvd101x** [comment]
Hi, I’ve been working on it and the current state seems approximate to what I would like.

In the following days I will migrate the logic to scroll and highlight the current output and clean up unused files from the CodeMirror5 example.

I think it’s looking good.

File: package.json [hunk 1]
```diff
@@ -26,6 +26,8 @@
   ],
   "dependencies": {
     "@babel/runtime": "^7.22.6",
+    "alpinejs": "^3.13.2",
```
Comment (josdejong) [inline package.json:29]: `alphinejs` and `codemirror` should be part listed in the package.json of the example, not of mathjs itself.

**Comment by josdejong** [comment]
O wow that definitely looks good 😎 , the `katex` rendering really makes a difference.

File: package.json (hunk 1 above)
Comment (dvd101x) [reply package.json]: It should be resolved now.

**Comment by dvd101x** [comment]
Hi

Over the weekend I will remove some unnecessary code and files.

**Comment by josdejong** [comment]
@dvd101x I see you made some more commits. What is the current status of the PR again?

**Comment by dvd101x** [comment]
Hi Jos. Thanks for the follow up. It's mostly done.

I would like for you to review how to integrate the build process into the project and webpage. I'm not familiar with how the webpage is built.

**Comment by dvd101x** [comment]
Hi Jos, thanks for your kind words and thorough review!

I will try to make my code more readable and comment it better.

1. Ok, thanks for catching this!
2. Ok
3. Ok
4. Ok
5. Definitely not mandatory to use Alpine. Somewhere I found the use of loadash and thought Alpine might help reduce the amount of code. I will remove it.

**Comment by dvd101x** [comment]
Hi, so far everything is done except Alpine, it might be more difficult than anticipated. Might take a while.

**Comment by josdejong** [comment]
🚀  thanks.

Do you want me to give that a try and see if removing Alphine and using plain HTML+JS works out OK?

**Comment by dvd101x** [comment]
Yes I would like to try. I'll report back

**Comment by dvd101x** [comment]
AlpineJS was removed successfully. I think the behavior is equivalent. Please review and let me know if something is missing.

**Comment by josdejong** [comment]
Thanks for the update David, that looks neat! There is indeed some extra logic now in `updateSelection` but it is easy to understand what is going on there.

When running the example now I get an error, and in the code I see that `processedResults` is undefined in `updateSelection`, and `previousSelectedExpressionIndex` as well. I guess these variables should be defined globally in `main.js`? There is also an unused variable `arrayOfResults`, can you remove that?

**Comment by josdejong** [comment]
Thanks! All looks good, merging your PR now.

**Comment by dvd101x** [comment]
Hi Jos, thanks for everything!

I was checking today the end result, but didn't find the link in [examples](https://mathjs.org/examples/index.html). Probably because the folders are different. I'm not familiar with the build process for the website, just wanted to let you know.

**Comment by josdejong** [comment]
That is a good point. The script to generate the examples pages on the website should be extended a bit I guess. This script is located in the `gh-pages` branch: https://github.com/josdejong/mathjs/blob/gh-pages/gulpfile.js#L166.

**Pull request merged**